etc/kernel-launchers/python/scripts/launch_ipykernel.py

```diff
@@ -357,14 +356,19 @@ def gateway_listener(sock, parent_pid):
                               stdin_port=ports[2], hb_port=ports[3], control_port=ports[4])
         if response_addr:
             gateway_socket = return_connection_info(connection_file, response_addr, lower_port, upper_port)
-            if gateway_socket:  # socket in use, start gateway listener
+            if gateway_socket:  # socket in use, start gateway listener thread
                 gateway_listener_process = Process(target=gateway_listener, args=(gateway_socket, os.getpid(),))
                 gateway_listener_process.start()
 
     # Initialize the kernel namespace for the given cluster type
     if cluster_type == 'spark' and spark_init_mode == 'none':
         cluster_type = 'none'
-    namespace = initialize_namespace(cluster_type=cluster_type)
+
+    spark_ns = None
+    if cluster_type == 'spark':
+        spark_ns = locals()
+
+    namespace = initialize_namespace(cluster_type=cluster_type, spark_ns=spark_ns)
```

Comment:
This isn't quite correct. You properly mutating the locals for spark, but you need to for dask as well (this isn't spark specific, it's just a namespace issue). You're also not clearing the namespace, so it's still polluted by everything in this initialization script.

I recommend the following:

- Change `initialize_namespace` to the following signature:

```python
def initialize_namespace(namespace, cluster_type="spark"):
    """Mutate the namespace in place to initialize it for the given cluster"
    if cluster_type == "spark":
       namespace.update(...)
    elif cluster_type == "dask":
        namespace.update(...)

- Add the following function:

```python
def start_ipython(namespace, cluster_type="spark", **kwargs):
    from IPython import embed_kernel

    # create an initial list of variables to clear
    # we do this without deleting to preserve the locals so that
    # initialize_namespace isn't affected by this mutation
    to_delete = [k for k in namespace if not k.startswith('__')]
    
    # initialize the namespace with the proper variables
    initialize_namespace(namespace, cluster_type=cluster_type)

    # delete the extraneous variables
    for k in to_delete:
        del namespace[k]

    # Start the kernel
    embed_kernel(local_ns=namespace, **kwargs)
```

- Replace the code here with the following:

```python
start_ipython(locals(), cluster_type=cluster_type, connection_file=connection_file, ip=ip)
```

You may need to update the lazy loading thread code to use local imports/references as the mutation of the external locals dict in `initialize_namespace` may break things. But for the dask side this should work at least.

Reply:
Thanks @jcrist.  So the for loop to remove extraneous variables ends up removing the various spark variables that, I thought, we wanted.  Are you sure we only want things in the namespace that start with `__`?

Reply:
Are you saying that the returned namespace has no `sc` in it, or that the proxy wrappers as written no longer work?

The code above assumes that `initialize_namespace` only adds variables to the existing namespace, and doesn't shadow any (that may already be in the `to_delete` list). If this isn't true, perhaps `initialize_namespace` should also return a list of variables added that should be kept? We do want to delete everything that doesn't start with `__`, as the user should be presented with a (mostly) empty namespace. `__things__` are part of python's internals and shouldn't be removed.

The `WaitingForSparkSessionToBeInitialized` will need to be rewritten to forward calls to take the passed in namespace and get the wrapped variable from in there rather than using the call to `globals()` in `__getattr__`.

Reply:
Sorry, I misread your updated code and forgot that the `sc` variables are getting added after the set of non-__ symbols have been built.  I'll give the above a shot.

Still barely following this discussion but have learned that python is way too lenient of a language - its kind of joke IMHO.

Thank you for your help!

Reply:
> but have learned that python is way too lenient of a language - its kind of joke IMHO.

To be fair, the use case here is quite uncommon - you're trying to mutate the locals of one frame directly from a different thread. I encourage not doing this (5-10 seconds isn't much, especially if the user is already planning on using spark anyway), but if you need to you're going to have to get into the runtime which complicates things. Messing with any language's runtime is bound to expose demons.

Reply:
@jcrist - please see the latest incantation.  I found that `WaitingForSparkSessionToBeInitialized` and `atexit` where getting removed from the namespace, but they're needed when the various spark variables are dereferenced.  Still unsure about `atexit`, but until it was kept, dereference of the spark variables would still blow out the stack.

As these changes are right now, everything works in yarn/spark.  I can't try Dask, but I suspect it will be fine - although those two symbols will be present.

Thanks again for all your help on this!